src/drivers/imu/bmi088/BMI088.hpp

```diff
@@ -0,0 +1,96 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2018 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <drivers/device/spi.h>
+#include <ecl/geo/geo.h>
+#include <lib/conversion/rotation.h>
+#include <lib/perf/perf_counter.h>
+#include <px4_getopt.h>
+#include <px4_work_queue/ScheduledWorkItem.hpp>
+
+#define DIR_READ                0x80
+#define DIR_WRITE               0x00
+
+//Soft-reset command Value
+#define BMI088_SOFT_RESET       0xB6
+
+#define BMI088_BUS_SPEED				10*1000*1000
+
+#define BMI088_TIMER_REDUCTION				200
+
+class BMI088 : public device::SPI
+{
+
+protected:
+
+    uint8_t         _whoami;    /** whoami result */
+
+    uint8_t         _register_wait;
+    uint64_t        _reset_wait;
+
+    enum Rotation       _rotation;
+
+    uint8_t         _checked_next;
+
+    /**
+   * Read a register from the BMI088
+   *
+   * @param       The register to read.
+   * @return      The value that was read.
+   */
+    virtual uint8_t         read_reg(unsigned reg); // This needs to be declared as virtual, because the
```

Comment:
Because the accelerometer part needs to do a dummy read before reading the actual data, it needs a different read_reg function than the gyro part, therefor I have declared this virtual, so that class pointers will call the correct function.

Perhaps I should just split this out into a read_reg in the gyro part and a separate one in the accel part, instead of having it here in the base class that is then overridden in the accel class.

Thoughts?